File: AllToDo-iOS/AllToDo/Services/Wasm/WasmManager.swift
import Foundation

final class WasmManager {
    static let shared = WasmManager()
    
    // Switch to Real Runtime
    private let runtime: WasmRuntime = WebViewWasmRuntime()
    private let session = URLSession(configuration: .default)
    
    // Updated IP
    private let advancedURL = URL(string: "http://175.194.163.56:8003/wasm/advanced")!
    private let versionURL = URL(string: "http://175.194.163.56:8003/wasm/version")!
    
    private init() {}
    
    struct VersionResponse: Codable {
        let version: String
    }
    
    func initialize(completion: @escaping (Bool) -> Void) {
        print("[WASM_STATUS] 🚀 Initializing WASM Manager...")
        
        Task {
            // 1. Storage Load
            var isLoaded = false
            if let (_, blobJson) = WasmStorage.shared.load() {
                if let bundle = try? JSONDecoder().decode(WasmBundle.self, from: blobJson),
                   let decrypted = try? WasmCrypto.shared.decrypt(bundle: bundle) {
                    do {
                        try await runtime.loadModule(decrypted)
                        print("[WASM_STATUS] ✅ Loaded stored WASM (Version: \(bundle.version))")
                        isLoaded = true
                    } catch {
                        print("[WASM_STATUS] ❌ Failed to load stored WASM: \(error)")
                    }
                }
            }
            
            if !isLoaded {
                 print("[WASM_STATUS] ℹ️ No valid stored WASM found. Loading Fallback & Checking Server...")
                 await loadFallback()
            }
            
            // Notify App is ready (using Stored or Fallback)
            completion(true)
            
            // 2. Check for Updates in Background (Unified Logic with Android)
            await checkForUpdate()
        }
    }
    
    private func checkForUpdate() async {
        print("[WASM_STATUS] 🔍 Checking for updates at: \(versionURL)...")
        do {
            let (data, _) = try await session.data(from: versionURL)
            let serverVer = try JSONDecoder().decode(VersionResponse.self, from: data).version
            
            let (storedVer, _) = WasmStorage.shared.load() ?? ("0.0.0", Data())
            
            if serverVer != storedVer {
                print("[WASM_STATUS] 🆕 New version found: \(serverVer) (Current: \(storedVer)). Downloading...")
                _ = await fetchAdvancedWasm()
            } else {
                print("[WASM_STATUS] ✅ Up to date (\(storedVer))")
            }
        } catch {
            print("[WASM_STATUS] ❌ Update check error: \(error.localizedDescription)")
        }
    }
    
    private func fetchAdvancedWasm() async -> Bool {
        print("[WASM_STATUS] 🔍 (1/3) Start Connection Check to \(advancedURL)...")
        do {
            var request = URLRequest(url: advancedURL)
            request.timeoutInterval = 5 // Short timeout for check
            
            let (data, response) = try await session.data(for: request)
            
            if let httpResponse = response as? HTTPURLResponse {
                if httpResponse.statusCode == 200 {
                    print("[WASM_STATUS] ✅ (2/3) Connection Successful! (Status: 200 OK)")
                } else {
                    print("[WASM_STATUS] ⚠️ (2/3) Connection Reached Server but Failed. Status: \(httpResponse.statusCode)")
                    return false
                }
            }
            
            let bundle = try JSONDecoder().decode(WasmBundle.self, from: data)
            let decrypted = try WasmCrypto.shared.decrypt(bundle: bundle)
            
            try await runtime.loadModule(decrypted)
            WasmStorage.shared.save(version: bundle.version, blobJson: data)
            
            print("[WASM_STATUS] 🎉 (3/3) Download & Load Complete. Version: \(bundle.version)")
            
            // [NEW] Self-Test
            await verifyWasm()
            
            return true
        } catch {
            print("[WASM_STATUS] ❌ Connection FAILED: \(error.localizedDescription)")
            // Detailed error for easier debugging
            print("[WASM_STATUS] 🛑 Error Details: \(error)")
            return false
        }
    }

    private func verifyWasm() async {
        print("[WASM_STATUS] 🧪 Starting WASM Self-Test (RDP + Clustering)...")
        do {
            // 1. RDP Test
            let rdpPoints: [Int32] = [0, 0, 100000, 100000, 200000, 200000]
            let rdpResult = try await runtime.compressTrajectory(rdpPoints, minDistMeters: 5.0, angleThreshDeg: 5.0)
            
            var rdpPassed = false
            if rdpResult.count == 4 {
                if rdpResult[0] == 0 && rdpResult[2] == 200000 {
                    rdpPassed = true
                }
            }
            
            // 2. Clustering Test
            let clusterPoints: [Int32] = [0, 0, 100, 100]
            // Safe to call now as Protocol is updated
            let clusterResult = try await runtime.clusterPoints(clusterPoints, cellSizeMeters: 100.0)
            
            var clusterPassed = false
            if clusterResult.count == 3 {
                 let count = clusterResult[2]
                 if count == 2 { clusterPassed = true }
            }
            
            if rdpPassed && clusterPassed {
                 print("[WASM_STATUS] ✅ WASM Self-Test PASSED (RDP + Clustering)")
            } else {
                 print("[WASM_STATUS] ❌ WASM Self-Test FAILED: RDP=\(rdpPassed), Cluster=\(clusterPassed)")
            }
        } catch {
             print("[WASM_STATUS] ❌ WASM Self-Test Error: \(error)")
        }
    }
    
    private func loadFallback() async {
        guard
            let url = Bundle.main.url(forResource: "fallback", withExtension: "wasm"),
            let data = try? Data(contentsOf: url)
        else {
            print("[WASM_STATUS] 🔥 CRITICAL: fallback.wasm not found in Resources!")
            return
        }
        
        do {
            try await runtime.loadModule(data)
            print("[WASM_STATUS] 📦 Loaded built-in Fallback WASM")
            RemoteLogger.info("Unavailable Server. Loaded Fallback WASM.")
        } catch {
            print("[WASM_STATUS] 🔥 Failed to load fallback WASM: \(error)")
            RemoteLogger.error("Failed to load Fallback WASM: \(error.localizedDescription)")
        }
    }
    
    public private(set) var lastErrorMessage: String? = nil

    func compress(points: [Int32]) async -> [Int32] {
        let start = Date()
        print("[WASM_STATUS] ⚡️ Executing WASM 'compressTrajectory' with \(points.count/2) points...")
        
        do {
            let result = try await runtime.compressTrajectory(points, minDistMeters: 3.0, angleThreshDeg: 10.0)
            
            let duration = Date().timeIntervalSince(start) * 1000
            let msg = "WASM Success: \(points.count/2) -> \(result.count/2) pts (\(String(format: "%.1f", duration))ms)"
            print("[WASM_STATUS] ✨ \(msg)")
            
            // RemoteLogger.info(msg) // Disabled to avoid spam with batch streaming. Parent handles it.
            lastErrorMessage = nil
            return result
        } catch {
            print("[WASM_STATUS] ❌ Execution Failed: \(error)")
            lastErrorMessage = error.localizedDescription
            RemoteLogger.error("WASM Exec Failed: \(error.localizedDescription)")
            return points // Return original on error
        }
    }
}

File: AllToDo-Android/app/src/main/java/kr/alltodo/wasm/WasmManager.kt
package kr.alltodo.wasm

import android.content.Context
import com.google.gson.Gson
import kr.alltodo.services.RemoteLogger
import okhttp3.OkHttpClient
import okhttp3.Request
import kotlinx.coroutines.*

import javax.inject.Inject
import javax.inject.Singleton
import dagger.hilt.android.qualifiers.ApplicationContext

@Singleton
class WasmManager @Inject constructor(
    @ApplicationContext private val context: Context
) {
    private val client = OkHttpClient()
    private val gson = Gson()
    private val storage = WasmStorage(context)
    private var runtime: WasmRuntime = WebViewWasmRuntime(context)
    private val advancedUrl = "http://175.194.163.56:8003/wasm/advanced"
    private val versionUrl = "http://175.194.163.56:8003/wasm/version"
    
    private val TAG = "[WASM_STATUS]"
    
    data class VersionResponse(val version: String)
    
    // [NEW] Debugging Properties for UI Overlay
    var lastErrorMessage: String? = null
        private set
        
    var onStatusUpdate: ((String) -> Unit)? = null

    private fun logStep(message: String) {
        val time = java.text.SimpleDateFormat("HH:mm:ss.SSS", java.util.Locale.getDefault()).format(java.util.Date())
        android.util.Log.e("WASM_LOG", "$time >>> WASM ($message)")
        System.out.println("$time >>> WASM ($message)")
    }

    fun initialize(onReady: (Boolean) -> Unit) {
        try {
            logStep(">>> WASM Loading Fallback...")
            val input = context.assets.open("fallback.wasm")
            val bytes = input.readBytes()
            runtime.loadModule(bytes)
            lastErrorMessage = null
            logStep(">>> WASM Fallback Loaded Success")
        } catch (e: Exception) {
            e.printStackTrace()
            lastErrorMessage = "Failed to load WASM"
            logStep(">>> WASM Fallback Load Failed")
        }
        // 비동기 방식으로 
        logStep(">>> WASM Server Connecting...")

        onReady(true)

        // 2. Check for Updates in Background
        CoroutineScope(Dispatchers.IO).launch {
            loadFromStorage()
            checkForUpdate()
        }
   }

    private suspend fun loadFromStorage() {
        storage.load()?.let { (_, blobJson) ->
            try {
                // A. Load into Candidate Runtime (Separate from Fallback)
                val candidateRuntime = WebViewWasmRuntime(context)
                
                val bundle = gson.fromJson(blobJson, WasmBundle::class.java)
                val decrypted = WasmCrypto.decrypt(bundle)
                
                withContext(Dispatchers.Main) {
                    candidateRuntime.loadModule(decrypted)
                }
                
                // B. Verify Candidate
                val isVerified = verifyWasm(candidateRuntime)
                
                // C. Hot-Swap if Verified
                if (isVerified) {
                    runtime = candidateRuntime
                    lastErrorMessage = null
                    onStatusUpdate?.invoke("Ready (Stored v${bundle.version})")
                    logStep(">>> WASM Loaded Stored v${bundle.version}")
                } else {
                    logStep(">>> WASM Stored Verification Failed")
                }
            } catch (e: Exception) {
                e.printStackTrace()
            }
        } ?: run {
            logStep(">>> WASM No stored WASM")
        }
    }
    
    // [NEW] Exposed method to replace Native Code
    suspend fun compress(points: List<Int>): List<Int> {
        val start = System.currentTimeMillis()
        
        val result = runtime.compressTrajectory(points, 3, 10)
        
        val duration = System.currentTimeMillis() - start
        val msg = "WASM Success: ${points.size/2} -> ${result.size/2} pts (${duration}ms)"
        lastErrorMessage = null
        withContext(Dispatchers.Main) {
            onStatusUpdate?.invoke("Comp: ${points.size/2}->${result.size/2} (${duration}ms)")
        }
        return result
    }

    // [NEW] Clustering Support
    suspend fun cluster(points: List<Int>, cellSizeMeters: Int): List<Int> {
        val start = System.currentTimeMillis()
        
        val result = runtime.clusterPoints(points, cellSizeMeters)
        
        val duration = System.currentTimeMillis() - start
        lastErrorMessage = null
        return result
    }

    private suspend fun checkForUpdate() {
        try {
            // A. Check Server Version
            logStep(">>> Checking Server Version...")
            val request = Request.Builder().url(versionUrl).build()
            val response = client.newCall(request).execute()
            if (!response.isSuccessful) {
                logStep(">>> Server Check Failed: ${response.code}")
                return
            }
            
            val serverVersion = gson.fromJson(response.body?.string(), VersionResponse::class.java).version
            val (storedVersion, _) = storage.load() ?: ("0.0.0" to "")
            
            if (serverVersion != storedVersion) {
                logStep(">>> New Version Found: $serverVersion (Current: $storedVersion)")
                logStep(">>> Downloading Advanced Module...")
                if (fetchAndLoadAdvanced()) {
                    RemoteLogger.info("WASM Upgraded to Version $serverVersion")
                    logStep(">>> Download & Load Success")
                } else {
                    logStep(">>> Download Failed")
                }
            } else {
                logStep(">>> Already Latest Version ($serverVersion)")
            }
        } catch (e: Exception) {
            e.printStackTrace()
            logStep(">>> Server Check Error: ${e.message}")
        }
    }

    private suspend fun fetchAndLoadAdvanced(): Boolean {
        return try {
            val request = Request.Builder().url(advancedUrl).build()
            val response = client.newCall(request).execute()
            
            if (!response.isSuccessful) {
                return false
            }
            
            val body = response.body?.string() ?: return false
            val bundle = gson.fromJson(body, WasmBundle::class.java)
            val decrypted = WasmCrypto.decrypt(bundle)
            
            // [Safety] Load into CANDIDATE Runtime first
            val candidateRuntime = WebViewWasmRuntime(context)
            
            withContext(Dispatchers.Main) {
                candidateRuntime.loadModule(decrypted)
            }
            
            // [Safety] Verify before Swap
            if (verifyWasm(candidateRuntime)) {
                runtime = candidateRuntime
                storage.save(bundle.version, body)
                logStep(">>> WASM Hot-Swapped to Server Version ${bundle.version}")
                return true
            } else {
                logStep(">>> Server WASM Verification Failed. Discarding.")
                return false
            }
        } catch (e: Exception) {
            e.printStackTrace()
            return false
        }
    }

    private suspend fun verifyWasm(targetRuntime: WasmRuntime): Boolean {
        return try {
            logStep("Running Self-Test...")
            // 1. RDP Test Case
            val rdpPoints = listOf(0, 0, 100000, 100000, 200000, 200000) 
            val rdpResult = targetRuntime.compressTrajectory(rdpPoints, 5, 5)
            
            val rdpPassed = if (rdpResult.size == 4) {
                 val p1 = "${rdpResult[0]},${rdpResult[1]}"
                 val p2 = "${rdpResult[2]},${rdpResult[3]}"
                 (p1 == "0,0" && p2 == "200000,200000")
            } else false
            
            // 2. Clustering Test Case
            val clusterPoints = listOf(0, 0, 100, 100)
            val clusterResult = targetRuntime.clusterPoints(clusterPoints, 100)
            
            val clusterPassed = if (clusterResult.size == 3) {
                val count = clusterResult[2]
                count == 2
            } else false

            if (rdpPassed && clusterPassed) {
                onStatusUpdate?.invoke("WASM Verified (RDP+Cluster)")
                logStep("Self-Test Passed")
                true
            } else {
                logStep("Self-Test Failed (RDP=$rdpPassed, Cluster=$clusterPassed)")
                false
            }
        } catch (e: Exception) {
            logStep("Self-Test Error: ${e.message}")
            false
        }
    }
}
